### DISCUSSION TITLE
Any support for "displayless" scenarios?

### DISCUSSION THREAD

**Discussion by Clockwork-Muse** [godotengine/godot-proposals#4134] [Engine Core]
One thing possible with some engines (Unreal, say) is the ability to work in "displayless" scenarios - that is, they're able to run _and render_ on a GPU, even when the system has no monitor/display (this is distinct from the "headless" scenario currently supported in godot, which cannot render at all).

There's a number of uses for something like this;
 - Game streaming
 - Render farms for video production or synthetic data generation for machine learning
 - CI/CD render output tests for engine development.

One advantage to having this enabled at the engine level is that it greatly simplifies the requirements on the deployment end; for instance, on Linux setting up a dummy display may not be possible in some hosted scenarios (insufficient permissions), or may create other conflicts.

**Comment by fire** [comment]
How do you expect to do this?

Unreal Engine has this concept of pixel streaming. As far as I can tell, it takes a system with GPU and sends it to a different application like viewing a webcam stream.

Also there's a similar entry in the GSOC 2022 prompt. https://github.com/godotengine/godot-proposals/issues/4064

## References

See https://docs.godotengine.org/en/stable/tutorials/export/exporting_for_dedicated_servers.html#headless-versus-server-binaries

and also https://github.com/godotengine/godot-proposals/issues/991

* https://github.com/Unity-Technologies/UnityRenderStreaming
* https://docs.unrealengine.com/4.27/en-US/SharingAndReleasing/PixelStreaming/

**Reply by Clockwork-Muse** [reply]
> How do you expect to do this?

In unreal, running any exported binary in "displayless" mode is done with a command line parameter, `-RenderOffScreen`.  This doesn't require the pixel streaming plugin at all (and also applies to the editor, I believe).  Either autodetecting no connected windows or ability to use a windowing system (eg, on linux detecting no connection to an X11 port), and/or the ability to pass an explicit "offscreen" parameter is fine.

> Also there's a similar entry in the GSOC 2022 prompt.

I'm not sure what you're referring to, but none of the existing listed prompts seem relevant, or would require the necessary backend work as a prerequisite.

> also https://github.com/godotengine/godot-proposals/issues/991

My read on that issue is that that dummy driver doesn't actually _render_ anything, which is what I need.

**Comment by Calinou** [comment]
See https://github.com/godotengine/godot-proposals/issues/1760, which would require off-screen rendering to be implemented. It's definitely planned at some point, but there is no ETA.

**Reply by Clockwork-Muse** [reply]
That issue specifically is calling out a software-based driver/renderer which is... not acceptable for anything other than the simple CI tests they're running.  Realistically, there would need to be a GPU-backed offscreen renderer for some of the other scenarios I mentioned (and should maybe be used for CI tests, too).

**Reply by Calinou** [reply]
Off-screen rendering can also be done with the GPU, but it will likely require access to specific OpenGL/Vulkan extensions. I think software off-screen rendering would piggy-back off the GPU implementation anyway.

To clarify, there are no plans to add a "true" software renderer to Godot – it's too inefficient. Instead, the GPU implementation of off-screen rendering would be run using a software Vulkan/OpenGL implementation in scenarios where no GPU is available. GPUs are very expensive to rent in servers, so we can't assume that everyone will have a GPU available there :slightly_smiling_face:

**Reply by fire** [reply]
I found this:

https://www.saschawillems.de/blog/2017/09/16/headless-vulkan-examples/

> The headless compute example fills an input buffer and runs a compute shader that calculates the fibonacci number for the current value and writes it back to the buffer:

> The headless rendering example renders a scene to a (non-visible) frame buffer attachment, copies the results from VRAM to host (using a host visible linear tiled image) and stores the result to disk.
